sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderTest.java

```diff
@@ -74,4 +79,35 @@ public void testNonDeterministicException() {
     assertThat(exception.toString(), containsString("Problem"));
     assertThat(exception.toString(), containsString("is not deterministic"));
   }
+
+  @Test
+  public void testTypeIsPreserved() throws Exception {
+    assertEquals(TypeDescriptor.of(Void.class), VoidCoder.of().getEncodedTypeDescriptor());
+  }
+
+  @Test
+  public void testGenericStandardCoderFallsBackToObject() throws Exception {
+    assertEquals(TypeDescriptor.of(Object.class), new Foo<String>().getEncodedTypeDescriptor());
+  }
+
+  private static class Foo<T> extends StandardCoder<T> {
+
+    @Override
+    public void encode(T value, OutputStream outStream, Coder.Context context)
+        throws CoderException, IOException {}
+
+    @Override
+    public T decode(InputStream inStream, Coder.Context context)
+        throws CoderException, IOException {
+      return null;
```

Comment:
Prefer to throw `UnsupportedOperationException` so it certainly fails if anything actually uses the methods that are not intended for test.